packages/autorest.go/src/gocodemodel/gocodemodel.ts

```diff
@@ -0,0 +1,1252 @@
+/*---------------------------------------------------------------------------------------------
+*  Copyright (c) Microsoft Corporation. All rights reserved.
+*  Licensed under the MIT License. See License.txt in the project root for license information.
+*--------------------------------------------------------------------------------------------*/
+
+// GoCodeModel contains a Go-specific abstraction over an OpenAPI (or other) description of REST endpoints.
+export interface GoCodeModel {
+  info: Info;
+
+  host?: string;
+
+  type: CodeModelType;
+
+  packageName: string;
+
+  options: Options;
+
+  // all of the struct model types to generate (models.go file)
+  models: Array<ModelType | PolymorphicType>;
+
+  // all of the const types to generate (constants.go file)
+  constants?: Array<ConstantType>;
+
+  // all of the operation groups (i.e. clients and their methods)
+  // no clients indicates a models-only build
+  clients?: Array<Client>;
+
+  // all of the parameter groups including the options types (options.go file)
+  paramGroups?: Array<StructType>;
+
+  // all of the response envelopes (responses.go file)
+  responseEnvelopes?: Array<ResponseEnvelope>;
+
+  // all of the interfaces for discriminated types (interfaces.go file)
+  interfaceTypes?: Array<InterfaceType>;
+
+  marshallingRequirements: MarshallingRequirements;
+}
+
+export type CodeModelType = 'azure-arm' | 'data-plane';
+
+// Info contains top-level info about the input source
+export interface Info {
+  title: string;
+}
+
+// Options contains global options set on the CodeModel.
+export interface Options {
+  headerText: string;
+
+  generateFakes: boolean;
+
+  injectSpans: boolean;
+
+  // NOTE: containingModule and module are mutually exclusive
+
+  // the module into which the package is being generated
+  containingModule?: string;
+
+  // the module being generated
+  module?: string;
+
+  moduleVersion?: string;
+
+  azcoreVersion?: string;
+}
+
+// MarshallingRequirements contains flags for required marshalling helpers
+export interface MarshallingRequirements {
+  generateDateHelper: boolean;
+
+  generateTimeRFC1123Helper: boolean;
+
+  generateTimeRFC3339Helper: boolean;
+
+  generateUnixTimeHelper: boolean;
+
+  generateXMLDictionaryUnmarshallingHelper: boolean;
+}
+
+// Struct describes a vanilla struct definition (pretty much exclusively used for parameter groups/options bag types)
+// UDTs that are sent/received are modeled as ModelType.
+export interface StructType {
+  name: string;
+
+  description?: string;
+
+  // there are only a few corner-cases where a struct has no fields
+  fields?: Array<StructField>;
```

Comment:
The only thing I was trying to indicate are the cases where it's legal for an array to contain no items, forcing you to programmatically handle that case.  That said, we aren't at present throwing any errors when an array that _shouldn't_ be empty is (maybe we should be).

Reply:
@richardpark-msft given your JS background, do you have any thoughts on this?

Reply:
An empty array sounds better - otherwise you have to check both it being null/undefined _and_ if it's empty. Might as well save the check.

From what you described it does sound like you're worried some places might not be checking, so it blowing up because of a null/undefined is better. Is that right?

Reply:
Not that it blows up, but whether or not it's expected/allowed to be empty.  The majority of arrays are always expected to contain at least one item, but we don't actually verify this at present.